### PR TITLE
 NO-ISSUE:  Add support for builtin hooks

### PR DESCRIPTION
**Pull request by ori-amizur**
Add support for builtin hooks in addition to the existing support of api
    hooks.
    The builtin hooks are actualy functions that perform specific task.
    This PR changes the implementation of podman-compose up and
    podman-compose down to use builtin hooks

### CODE REVIEW

File: internal/agent/device/hook/actions.go [hunk 1]
```diff
@@ -259,3 +367,53 @@ func (e *executableActionHook) OnChange(ctx context.Context, path string) error
...
+		return nil
+	}
+	_, err := os.Stat(path)
+	if !os.IsNotExist(err) {
+		return logAndReturn(log, "podman-compose down failed for existing path %s with exit code %d: %s", path, exitCode, stderr)
+	}
+	psStdout, psStderr, psExitCode := exec.ExecuteWithContext(ctx, "podman", "ps", "-a", "--format=json")
+	if psExitCode != 0 {
+		log.Errorf("podman ps --all failed: %s", psStderr)
+		return logAndReturn(log, "podman-compose down failed for path %s with exit code %d: %s", path, exitCode, stderr)
```
Comment (hexfusion) [inline internal/agent/device/hook/actions.go:402]: instead of logAndReturn func can we return the error then log it?

**Review by hexfusion — COMMENTED** [review]
Thanks Ori I love where this is heading, left a few notes.

File: internal/agent/device/hook/actions.go (hunk 1 above)
Comment (ori-amizur) [reply internal/agent/device/hook/actions.go:402]: OK

File: internal/agent/device/hook/actions.go [hunk 2]
```diff
@@ -259,3 +367,53 @@ func (e *executableActionHook) OnChange(ctx context.Context, path string) error
...
+	}
+	type psRecord struct {
+		Labels map[string]string `json:"Labels"`
+	}
+	var psRecords []psRecord
+	if err = json.Unmarshal([]byte(psStdout), &psRecords); err != nil {
+		log.WithError(err).Errorf("json unmarshal failed:")
```
[the remark this answers is not in the document]
Comment (ori-amizur) [reply internal/agent/device/hook/actions.go:409]: This also uses log.Errof directly.  It is just a way to add the error in the message

File: internal/agent/device/hook/manager.go [hunk 3]
```diff
@@ -72,68 +84,41 @@ func NewManager(exec executer.Executer, log *log.PrefixLogger) Manager {
...
-		spec.Systemd.Unit.Operations,
-		m.exec,
-		actionTimeout,
-		m.log), nil
+	return &hookDefinition{
+		name:        lo.FromPtr(hookSpec.Name),
+		description: lo.FromPtr(hookSpec.Description),
+		actionHooks: actionHooks,
+		ops:         lo.FromPtr(hookSpec.OnFile),
+		path:        lo.FromPtr(hookSpec.Path),
```
[the remark this answers is not in the document]
Comment (ori-amizur) [reply internal/agent/device/hook/manager.go:97]: There is another usage in this file for this package

File: internal/agent/device/hook/actions.go [hunk 4]
```diff
@@ -259,3 +367,53 @@ func (e *executableActionHook) OnChange(ctx context.Context, path string) error
...
+
+	_, stderr, exitCode := exec.ExecuteWithContext(ctx, "podman-compose", "-f", path, "down")
+	if exitCode == 0 {
+		return nil
+	}
+	_, err := os.Stat(path)
+	if !os.IsNotExist(err) {
+		return logAndReturn(log, "podman-compose down failed for existing path %s with exit code %d: %s", path, exitCode, stderr)
+	}
+	psStdout, psStderr, psExitCode := exec.ExecuteWithContext(ctx, "podman", "ps", "-a", "--format=json")
```
[the remark this answers is not in the document]
Comment (ori-amizur) [reply internal/agent/device/hook/actions.go:399]: OK

File: internal/agent/device/hook/actions.go [hunk 5]
```diff
@@ -259,3 +367,53 @@ func (e *executableActionHook) OnChange(ctx context.Context, path string) error
 
 	return nil
 }
+
+func logAndReturn(log *log.PrefixLogger, format string, args ...any) error {
+	log.Errorf(format, args...)
+	return fmt.Errorf(format, args...)
+}
+
+func podmanComposeUp(ctx context.Context, path string, exec executer.Executer, log *log.PrefixLogger) error {
```
[the remark this answers is not in the document]
Comment (ori-amizur) [reply internal/agent/device/hook/actions.go:376]: OK

File: internal/agent/device/hook/default_hooks.go [hunk 6]
```diff
@@ -19,55 +19,52 @@ func marshalExecutable(run string, envVars *[]string, workDir string, timeout st
...
-			Description: util.StrToPtr("Bring down a multi-container system based on the provided YAML podman-compose definition file"),
-			OnFile:      &[]v1alpha1.FileOperation{v1alpha1.FileOperationUpdate, v1alpha1.FileOperationRemove},
-			Actions: []v1alpha1.HookAction{
-				marshalExecutable(`output=$(podman-compose -f {{ .FilePath }} down 2>&1); exit_code=$? ; \
-										if [[ $exit_code -eq 0 ]] || ( [[ ! -f {{ .FilePath }} ]] && \
-                                            ! ( podman ps -a --format=json | jq -r '.[] | .Labels."com.docker.compose.project.config_files"' | grep -q  -E '^{{ .FilePath }}$' )) ; then  \ 
-											exit 0; \
-										fi ; \
-										echo "$output" 1>&2 && exit $exit_code`, nil,
-					"/var/run/flightctl/compose", "1m"),
```
Comment (hexfusion) [inline internal/agent/device/hook/default_hooks.go:61]: :+1:

**Review by hexfusion — APPROVED** [review]

**Pull request merged**